File: backend/src/handlers/getSlots.ts
/**
 * getSlots.ts
 * ---------------------------------------------------------------------------
 * GET /slots
 * Returns every appointment slot in the SlotsTable.
 * Public read endpoint — no PII is exposed.
 * ---------------------------------------------------------------------------
 */

import type { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
import { ScanCommand } from '@aws-sdk/lib-dynamodb';
import { jsonResponse, errorResponse } from '../utils/response.js';
import { dynamoClient } from '../utils/dynamo.js';

/**
 * Lambda handler for listing all appointment slots.
 *
 * @param event - API Gateway proxy event.
 * @returns 200 with a list of slots, or 500 on unexpected error.
 */
export async function lambdaHandler(event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> {
  const tableName = process.env.SLOTS_TABLE;
  if (!tableName) {
    return errorResponse(500, 'Missing SLOTS_TABLE configuration', undefined, event);
  }

  try {
    const items: Array<Record<string, unknown>> = [];
    let exclusiveStartKey: Record<string, unknown> | undefined;

    // Paginate through the full table (fine for capstone-scale data).
    do {
      const result = await dynamoClient.send(
        new ScanCommand({
          TableName: tableName,
          ExclusiveStartKey: exclusiveStartKey,
        }),
      );

      items.push(...(result.Items ?? []));
      exclusiveStartKey = result.LastEvaluatedKey;
    } while (exclusiveStartKey);

    // Keep newest slots first, matching the frontend's unshift() behaviour.
    items.sort((a, b) => {
      const aCreated = String(a.createdAt ?? '');
      const bCreated = String(b.createdAt ?? '');
      return bCreated.localeCompare(aCreated);
    });

    return jsonResponse(200, items, event);
  } catch {
    // Do NOT leak internal error details to the client.
    return errorResponse(500, 'Failed to fetch slots', undefined, event);
  }
}

File: backend/local_handlers/get_slots.py
"""
get_slots.py
---------------------------------------------------------------------------
GET /slots
Returns every appointment slot in the SlotsTable.
Public read endpoint — no PII is exposed.
---------------------------------------------------------------------------
"""
import json
import os
import boto3
from decimal import Decimal

dynamodb = boto3.resource("dynamodb")
slots_table = dynamodb.Table(os.environ["SLOTS_TABLE"])

CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Headers": "Content-Type,Authorization",
    "Access-Control-Allow-Methods": "GET,OPTIONS"
}


def decimal_default(obj):
    if isinstance(obj, Decimal):
        return int(obj) if obj % 1 == 0 else float(obj)
    raise TypeError


def lambda_handler(event, context):
    try:
        response = slots_table.scan(
            ProjectionExpression="id, hospitalName, area, department, doctorName, "
                                 "doctorTitle, #dt, availableSeats, totalSeats, #s, consultationFee",
            ExpressionAttributeNames={"#dt": "dateTime", "#s": "status"}
        )
        items = response.get("Items", [])

        # Paginate through the full table (fine for capstone-scale data)
        while "LastEvaluatedKey" in response:
            response = slots_table.scan(ExclusiveStartKey=response["LastEvaluatedKey"])
            items.extend(response.get("Items", []))

        # Keep newest slots first, matching the mock frontend's unshift() behaviour
        items.sort(key=lambda s: s.get("createdAt", ""), reverse=True)

        return {
            "statusCode": 200,
            "headers": CORS_HEADERS,
            "body": json.dumps(items, default=decimal_default)
        }
    except Exception:
        # Do NOT leak internal error details to the client.
        return {
            "statusCode": 500,
            "headers": CORS_HEADERS,
            "body": json.dumps({"error": "Failed to fetch slots"})
        }